src/freenet/clients/fcp/FCPPluginClient.java

```diff
@@ -0,0 +1,1190 @@
+     *             This FCPPluginClient <b>should be</b> considered as dead once this happens, you
+     *             should then discard it and obtain a fresh one.
+     * @throws InterruptedException
+     *             If another thread called {@link Thread#interrupt()} upon the thread which you
+     *             used to execute this function.<br>
+     *             This is a shutdown mechanism: You can use it to abort a call to this function
+     *             which is waiting for the timeout to expire.<br><br>
+     * @see FCPPluginClient#synchronousSends
+     *          An overview of how synchronous sends and especially their threading work internally
+     *          is provided at the map which stores them.
+     * @see #send(SendDirection, FCPPluginMessage)
+     *          The non-blocking, asynchronous send() should be used instead of this whenever
+     *          possible.
+     */
+    public FCPPluginMessage sendSynchronous(SendDirection direction, FCPPluginMessage message,
+            long timeoutNanoSeconds)
+                throws IOException, InterruptedException {
+        
+        if(message.isReplyMessage()) {
+            throw new IllegalArgumentException("sendSynchronous() cannot send reply messages: " +
+                "If it did send a reply message, it would not get another reply back. " +
+                "But a reply is needed for sendSynchronous() to determine when to return.");
+        }
+        
+        assert(timeoutNanoSeconds > 0) : "Timeout should not be negative";
+        
+        assert(timeoutNanoSeconds <= TimeUnit.MINUTES.toNanos(1))
+            : "Please use sane timeouts to prevent thread congestion";
+        
+        
+        synchronousSendsLock.writeLock().lock();
+        try {
+            final Condition completionSignal = synchronousSendsLock.writeLock().newCondition();
+            final SynchronousSend synchronousSend = new SynchronousSend(completionSignal);
+            
+            // An assert() instead of a throwing is fine:
+            // - The constructor of FCPPluginMessage which we tell the user to use in the JavaDoc
+            //   does generate a random identifier, so collisions will only happen if the user
+            //   ignores the JavaDoc or changes the constructor.
+            // - If the assert is not true, then the following put() will replace the old
+            //   SynchronousSend, so its Condition will never get signaled, and its
+            //   thread waiting in sendSynchronous() will timeout safely. It IS possible that this
+            //   thread will then get a reply which does not belong to it. But the wrong reply will
+            //   only affect the caller, the FCPPluginClient will keep working fine, especially
+            //   no threads will become stalled for ever. As the caller is at fault for the issue,
+            //   it is fine if he breaks his own stuff :) The JavaDoc also documents this.
+            
+            as
```

Comment:
Nice catch!
Fixed by cdce839369b4d1ede9d8430790910230780d3b7e by moving the function to the message dispatcher class.
(It is still public so it can override toString(), which it was renamed to since there was no toString() there yet.)